src/timeout.ts

```diff
@@ -108,3 +108,131 @@ export class Timeout extends Promise<never> {
     );
   }
 }
+
+/** @internal */
+export type TimeoutContextOptions = {
+  timeoutMS?: number;
+  serverSelectionTimeoutMS: number;
+  waitQueueTimeoutMS: number;
+  socketTimeoutMS?: number;
+};
+
+/** @internal */
+export abstract class TimeoutContext {
+  private _csotEnabled: boolean;
+  clearServerSelectionTimeout: boolean;
+  clearConnectionCheckoutTimeout: boolean;
+
+  constructor(options: TimeoutContextOptions) {
+    this._csotEnabled = options.timeoutMS != null;
+    this.clearServerSelectionTimeout = true;
+    this.clearConnectionCheckoutTimeout = true;
+  }
+
+  static create(options: TimeoutContextOptions): TimeoutContext {
+    if (options.timeoutMS != null) return new CSOTTimeoutContext(options);
+    else return new LegacyTimeoutContext(options);
+  }
+
+  get serverSelectionTimeout(): Timeout | null {
+    return null;
+  }
+
+  get connectionCheckoutTimeout(): Timeout | null {
+    return null;
+  }
+
+  csotEnabled(): this is CSOTTimeoutContext {
+    return this._csotEnabled;
+  }
+}
+
+/** @internal */
+export class CSOTTimeoutContext extends TimeoutContext {
+  timeoutMS: number;
+  serverSelectionTimeoutMS: number;
+  socketTimeoutMS?: number;
+
+  private _maxTimeMS?: number;
+
+  private _serverSelectionTimeout?: Timeout | null;
+  private _connectionCheckoutTimeout?: Timeout | null;
```

Comment:
Just my thought - it might be simpler to never cache promises, at least until CSOT's implementation is complete and we can benchmark the difference between caching and not caching.  If caching a promise doesn't provide a noticable performance increase, then why add the complexity? (this also simplifies the interface, because now places that use timeouts can always clear the timeout.)

ex:
```typescript
override get serverSelectionTimeout(): Timeout | null {
    const timeout = this.serverSelectionTimeoutMS !== 0 &&
      csotMin(this.timeoutMS, this.serverSelectionTimeoutMS)

    return timeout ? Timeout.expires(timeout) : null;
}
```

as another minor modification, the Timeout class does nothing if the timeout parameter is 0, so we can actually always a return a timeout, right?
```typescript
override get serverSelectionTimeout(): Timeout {
    const timeout = this.serverSelectionTimeoutMS !== 0 &&
      csotMin(this.timeoutMS, this.serverSelectionTimeoutMS);

    return timeout ? Timeout.expires(timeout ?? 0);
}

// server selection, for example:
// unified code path, regardless of whether or not a timeout is being applied
try {
	timeout.throwIfExpired(); // <- as a question: is this really required? If the timeout has expired, wouldn't promise.race immediately reject with the timeout error?
	return Promise.race([serverPromise, timeout]);
} catch (error) {
	...
} finally {
  timeout.clear();
}
```

Reply:
Re the second comment, true, the Timeout constructor doesn't create an actual NodeJS.Timeout unless the duration passed in is greater than zero, so I can refactor that.

The reason I had done the caching approach was in anticipation of the socket timeout and maxTimeMS portions of the work. That's where I intend to introduce the overall timeout and use that for the the remaining sections of blocking work.

Reply:
Were you intending to refactor the timeout context to always return timeouts (second point)?  Or did you intended to leave it as-is?

Reply:
Looking more into the refactor, it would end up with us creating a lot more promises on hot paths than we'd need. For one, we'd create the 0 timeout, and then in both connection checkout and server selection, we'd end up creating a new promise for the race with the timeout, so I think it's better for us to just return null and eat the complexity which that requires in `TimeoutContext`